storage/pebble/execution_data_tracker.go

```diff
@@ -0,0 +1,406 @@
+	if err != nil {
+		return fmt.Errorf("failed to set fulfilled height value: %w", err)
+	}
+
+	return nil
+}
+
+// GetFulfilledHeight returns the current fulfilled height.
+//
+// No errors are expected during normal operation.
+func (s *ExecutionDataTracker) GetFulfilledHeight() (uint64, error) {
+	var fulfilledHeight uint64
+
+	err := operation.RetrieveTrackerFulfilledHeight(&fulfilledHeight)(s.db)
+	if err != nil {
+		return 0, err
+	}
+
+	return fulfilledHeight, nil
+}
+
+// setPrunedHeight updates the current pruned height.
+//
+// No errors are expected during normal operation.
+func (s *ExecutionDataTracker) setPrunedHeight(height uint64) error {
+	err := operation.UpdateTrackerPrunedHeight(height)(s.db)
+	if err != nil {
+		return fmt.Errorf("failed to set pruned height value: %w", err)
+	}
+
+	return nil
+}
+
+// GetPrunedHeight returns the current pruned height.
+//
+// No errors are expected during normal operation.
+func (s *ExecutionDataTracker) GetPrunedHeight() (uint64, error) {
+	var prunedHeight uint64
+
+	err := operation.RetrieveTrackerPrunedHeight(&prunedHeight)(s.db)
+	if err != nil {
+		return 0, err
+	}
+
+	return prunedHeight, nil
+}
+
+// trackBlob tracks a single blob CID at the specified block height.
+// This method first inserts a record of the blob at the specified block height.
+// It then checks if the current block height is greater than the previously recorded
+// latest height for this CID. If so, it updates the latest height to the current block height.
+//
+// Parameters:
+// - blockHeight: The height at which the blob was observed.
+// - c: The CID of the blob to be tracked.
+//
+// No errors are expected during normal operation.
+func (s *ExecutionDataTracker) trackBlob(blockHeight uint64, c cid.Cid) error {
+	err := operation.InsertBlob(blockHeight, c)(s.db)
+	if err != nil {
+		return fmt.Errorf("failed to add blob record: %w", err)
+	}
+
+	var latestHeight uint64
+	err = operation.RetrieveTrackerLatestHeight(c, &latestHeight)(s.db)
+	if err != nil {
+		if !errors.Is(err, storage.ErrNotFound) {
+			return fmt.Errorf("failed to get latest height: %w", err)
+		}
+	} else {
+		// don't update the latest height if there is already a higher block height containing this blob
+		if latestHeight >= blockHeight {
+			return nil
+		}
+	}
+
+	err = operation.UpsertTrackerLatestHeight(c, blockHeight)(s.db)
+	if err != nil {
+		return fmt.Errorf("failed to set latest height value: %w", err)
+	}
+
+	return nil
+}
+
+// trackBlobs tracks multiple blob CIDs at the specified block height.
+//
+// Parameters:
+// - blockHeight: The height at which the blobs were observed.
+// - cids: The CIDs of the blobs to be tracked.
+//
+// No errors are expected during normal operation.
+func (s *ExecutionDataTracker) trackBlobs(blockHeight uint64, cids ...cid.Cid) error {
+	if len(cids) > 0 {
+		for _, c := range cids {
+			if err := s.trackBlob(blockHeight, c); err != nil {
```

Comment:
ideally, we'd use a single batch for all blob writes for a block. that way if there's a failure or crash, the state is consistent

Reply:
[34e04ee](https://github.com/onflow/flow-go/pull/6277/commits/34e04eec57c1bfa33cf840f8b107cc507bcf6c24)